### ISSUE TITLE
[Feature request] Skip token entry on headless mode by passing token as URL parameter

### ISSUE THREAD

**Issue by tylerburleigh** [expfactory/expfactory#92]
### New behavior

If a participant accesses an experiment with the token ID as part of the URL, for example...
`https://expfactory.com/?token=63abe986-bd4a-4b70-ad2b-2ed58d1a4021`

Then they skip the token entry / login prompt and go straight to the experiment.

### Purpose of feature

This will eliminate a step in the participant onboarding process, reducing participant fatigue and likely improving compliance / completion rates. It's an ideal behavior in experimental settings like MTurk, where a participant's worker ID could be passed to expfactory as soon as they access the study URL from MTurk.

**Comment by vsoch** [comment]
The token entry is a standard form, e.g.:

https://github.com/expfactory/expfactory/blob/daf9aa4fc9b0f0b2f22c8067d4822b25cfc5f2b9/expfactory/forms.py#L51

And it's submit with POST, which is why the function accepts POST. Having a url parameter like that would be akin to a GET with the variable.. The view right now only allows a POST (typical for forms) https://github.com/expfactory/expfactory/blob/daf9aa4fc9b0f0b2f22c8067d4822b25cfc5f2b9/expfactory/views/headless.py#L64 so what you could first try is to change that to a list to accept both POST (from the form on the page) and GET (from the url). Do you want to give that a go?

Another consideration is the difference between POST and GET, period. POST is more secure because it's not openly viewable in the url, Read about the two here --> https://security.stackexchange.com/questions/33837/get-vs-post-which-is-more-secure and 
https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/Sending_and_retrieving_form_data

Basically, allowing the token (which is supposed to be secured for a participant) to be less secure is not something I'd advise. Given that the user just needs to copy paste it once and it's stored in a session, this isn't a huge deal with respect to effort.

You could also add a global password to the entire server, or some other custom portal, and then forget about headless and let the user take and finish experiments with the server managing their id. But that would be much harder for you to know who is who.

**Comment by tylerburleigh** [comment]
Thanks for pointing me in the right direction!

I followed your advice and made the changes within `expfactory/views/headless.py`, as here:

https://github.com/tylerburleigh/expfactory/commit/6033d91d99e46e1e0c35c958500fbfc6ba44c7ce

> Given that the user just needs to copy paste it once and it's stored in a session, this isn't a huge deal with respect to effort.

I think for some demographics it is a big deal. Individuals with lower IQ or less tech skills may find the extra step difficult. Some may not know how to copy and paste. The default tokens are long and can sometimes include ambiguous characters (e.g., `l` vs `1` vs `I`), so the act of typing it out can become very difficult.

Even for the average user, every additional step in the process of participating is another opportunity for drop-out. The cost of drop-out is huge for longitudinal studies (like the one I am working on), so every chance to optimize user experience is worth the effort.

>Another consideration is the difference between POST and GET, period. POST is more secure because it's not openly viewable in the url... Basically, allowing the token (which is supposed to be secured for a participant) to be less secure is not something I'd advise.

I think there are many designs where GET parameters are secure enough and the benefits of a better UX  justify their use. For example, in my study the `tokens` will be generated outside of expfactory and stored in a `master data file` along with the participant's `private id`. In this design, only the researcher who has access to the `master data file` can link a participant's expfactory data back to their identity. The `token` is therefore an `anonymized` identifier -- basically meaningless by itself.

**Comment by vsoch** [comment]
> I think for some demographics it is a big deal. Individuals with lower IQ or less tech skills may find the extra step difficult.

This demographic perhaps isn't suited for not supervised, headless runs?

> Some may not know how to copy and paste. The default tokens are long and can sometimes include ambiguous characters (e.g., l vs 1 vs I), so the act of typing it out can become very difficult.

If the demographic has this trouble, it might be administered in person, or by a proctor. I used to run a 3 hour battery with (highly intelligent) undergrads, and even with this population it was essential to do in person.

> Even for the average user, every additional step in the process of participating is another opportunity for drop-out. The cost of drop-out is huge for longitudinal studies (like the one I am working on), so every chance to optimize user experience is worth the effort.

Yeah, definitely agree!

> I think there are many designs where GET parameters are secure enough and the benefits of a better UX justify their use. For example, in my study the tokens will be generated outside of expfactory and stored in a master data file along with the participant's private id. In this design, only the researcher who has access to the master data file can link a participant's expfactory data back to their identity. The token is therefore an anonymized identifier -- basically meaningless by itself.

Sure, if you think this is secure, your change should do the trick! Overall, having the (hard to remember) identifiers, via POST, ensures moreso that someone can't enter a known identifier (or discover one) and corrupt the data. I agree with you on all fronts about participation, so it comes down to the risks that you are willing to take wrt security. You can definitely choose to favor participation, and so I can offer help to change the code to do this, although I'm not convinced it should be a default. Maybe there is a better solution that ensures security and ease of use, we can leave this open for more feedback and discussion.

**Comment by vsoch** [comment]
Here is a little more on secure forms, if it helps --> https://flask-wtf.readthedocs.io/en/latest/form.html?highlight=GET#module-flask_wtf

**Comment by tfeige91** [comment]
Hi @tylerburleigh,

I am very interested in your solution as we are planning to conduct some cognitive tests with a not so tech affine cohort.
Unfortunately the link in your answer is not working anymore. 
Can you upload or send me your modified version? 

Thanks in advance!

**Comment by vsoch** [comment]
I think we added support for this.
